### ISSUE TITLE
colorama breaks randomly on Windows 10 and Python 3.6

### ISSUE THREAD

**Issue by segevfiner** [pytest-dev/pytest#2465]
By the very same issue that causes https://github.com/pytest-dev/py/issues/103 the handles used by colorama are getting closed by Pytest's `FDCapture`. Though it only manifests randomly.

colorama ignores the errors though and all colorama features just randomly stop working (e.g. Colors stop working). Adding error handling to `colorama.winterm` reveals the very same exception. Essentially colorama caches the std handles in `colorama.win32.handles`. Which then get closed by Pytest's `FDCapture` calling `dup2`.

This probably affects other Python modules or anything else loaded into Pytest's process (Native libraries) that caches the stdio handles. And the underlying issue might have other affects besides this. I can think of `pyreadline`, `win-unicode-console` and `click` for example being broken by this but I haven't tested.

I am running this with `pytest -v` and randomly the green PASSED text stops being colored:
```python
import time
import random
import pytest


@pytest.mark.parametrize("i", range(1000))
def test_random_sleep(i, capfd):
    time.sleep(random.random() * 0.2)
```

**EDIT:** You might not be able to reproduce this because Pytest will break by https://github.com/pytest-dev/py/issues/103 first. It still happens with `PYTHONLEGACYWINDOWSSTDIO` since it's unrelated to WindowsConsoleIO breaking, but is still caused by the same cause. It might also happen in older Python 3 versions. Might be related to the VS/CRT 2015 switch.

## **EDIT:** Possible Fixes (At least ones I can come up with)
1. colorama can stop caching the console handles. Not sure if this has performance implications.
2. colorama can `DuplicateHandle` the console handles.
3. We can overwrite colorama's handles with duplicates our self.

## See
* https://github.com/pytest-dev/pytest/pull/2462
* https://github.com/pytest-dev/py/issues/103
* **EDIT:** https://bugs.python.org/issue30555 - A similar issue with WindowsConsoleIO in CPython
* https://github.com/tartley/colorama/issues/119

## System Info
```
Package    Version                         Location
---------- ------------------------------- -------------------------
certifi    2017.4.17
chardet    3.0.3
colorama   0.3.9
hypothesis 3.11.1
idna       2.5
mock       2.0.0
nose       1.3.7
pbr        3.0.1
pip        9.0.1
py         1.4.33
pytest     3.1.2.dev14+g05796be2.d20170602 c:\users\segev\prj\pytest
requests   2.17.3
setuptools 36.0.1
six        1.10.0
urllib3    1.21.1
wheel      0.29.0
```

Windows 10.0.14393 x64 with Python 3.6.1

**EDIT:** Also reproduces for me on Python 3.5.3.

Thanks for submitting an issue!

Here's a quick checklist in what to include:

- [X] Include a detailed description of the bug or suggestion
- [X] `pip list` of the virtual environment you are using
- [X] pytest and operating system versions
- [X] Minimal example if possible

**Comment by nicoddemus** [comment]
Fixed in #2462

**Comment by segevfiner** [comment]
@nicoddemus Actually #2462 fixes https://github.com/pytest-dev/py/issues/103. This one is a seperate issue since colorama is managing it's own Windows console handles and gets broken independently of what #2462 fixes. I submitted a PR to colorama which should fix this one too: https://github.com/tartley/colorama/pull/131.

This one even happens on Python 3.5 which doesn't even have WindowsConsoleIO since the console handle closing is likely a change in the crt.

**Comment by nicoddemus** [comment]
Oops sorry, my mistake! 😅

**Comment by nicoddemus** [comment]
@segevfiner actually is there anything else to do on pytest's side? Seems like the ball is in `colorama`'s court here.

**Comment by segevfiner** [comment]
@nicoddemus Not unless someone can come up with a workaround that avoids this issue in the first place. This sort of issue will repeat itself in many Python libraries which cache the std handles.

**Comment by nicoddemus** [comment]
OK thanks, let's keep it open to track it at least.

**Comment by nicoddemus** [comment]
With colorama's `0.4.0` release this should now be fixed. @segevfiner can we close this?

**Comment by segevfiner** [comment]
> With colorama's `0.4.0` release this should now be fixed. @segevfiner can we close this?

Yep. This is fixed with `colorama` 0.4.0.
